exporter/signalfxexporter/dimensions/dimclient.go

```diff
@@ -0,0 +1,340 @@
+// won't block. If the buffer is full, the dim update will be dropped.
+func (dc *DimensionClient) acceptDimension(dimUpdate *DimensionUpdate) error {
+	dc.Lock()
+	defer dc.Unlock()
+
+	if delayedDimUpdate := dc.delayedSet[dimUpdate.Key()]; delayedDimUpdate != nil {
+		if !reflect.DeepEqual(delayedDimUpdate, dimUpdate) {
+			dc.TotalFlappyUpdates++
+		}
+
+		// Merge the latest updates into existing one.
+		delayedDimUpdate.Properties = mergeProperties(delayedDimUpdate.Properties, dimUpdate.Properties)
+		delayedDimUpdate.Tags = mergeTags(delayedDimUpdate.Tags, dimUpdate.Tags)
+	} else {
+		atomic.AddInt64(&dc.DimensionsCurrentlyDelayed, int64(1))
+
+		dc.delayedSet[dimUpdate.Key()] = dimUpdate
+		select {
+		case dc.delayedQueue <- &queuedDimension{
+			DimensionUpdate: dimUpdate,
+			TimeToSend:      dc.now().Add(dc.sendDelay),
+		}:
+			break
+		default:
+			dc.TotalDimensionsDropped++
+			atomic.AddInt64(&dc.DimensionsCurrentlyDelayed, int64(-1))
+			return errors.New("dropped dimension update, propertiesMaxBuffered exceeded")
+		}
+	}
+
+	return nil
+}
+
+// mergeProperties merges 2 or more maps of properties. This method gives
+// precedence to values of properties in later maps. i.e., if more than one
+// map has the same key, the last value seen will be the effective value in
+// the output.
+func mergeProperties(propMaps ...map[string]*string) map[string]*string {
+	out := map[string]*string{}
+	for _, propMap := range propMaps {
+		for k, v := range propMap {
+			out[k] = v
+		}
+	}
+	return out
+}
+
+// mergeTags merges 2 or more sets of tags. This method gives precedence to
+// tags seen in later sets. i.e., if more than one set has the same tag, the
+// last value seen will be the effective value in the output.
+func mergeTags(tagSets ...map[string]bool) map[string]bool {
+	out := map[string]bool{}
+	for _, tagSet := range tagSets {
+		for k, v := range tagSet {
+			out[k] = v
+		}
+	}
+	return out
+}
+
+func (dc *DimensionClient) processQueue() {
+	for {
+		select {
+		case <-dc.ctx.Done():
+			return
+		case delayedDimUpdate := <-dc.delayedQueue:
+			now := dc.now()
+			if now.Before(delayedDimUpdate.TimeToSend) {
+				// dims are always in the channel in order of TimeToSend
+				time.Sleep(delayedDimUpdate.TimeToSend.Sub(now))
+			}
+
+			atomic.AddInt64(&dc.DimensionsCurrentlyDelayed, int64(-1))
+
+			dc.Lock()
+			delete(dc.delayedSet, delayedDimUpdate.Key())
+			dc.Unlock()
+
+			if err := dc.setPropertiesOnDimension(delayedDimUpdate.DimensionUpdate); err != nil {
+				dc.logger.Error(
+					"Could not send dimension update",
+					zap.Error(err),
+					zap.String("dimensionUpdate", delayedDimUpdate.String()),
+				)
+			}
+		}
+	}
+}
+
+// setPropertiesOnDimension will set custom properties on a specific dimension value.
+func (dc *DimensionClient) setPropertiesOnDimension(dimUpdate *DimensionUpdate) error {
+	var (
+		req *http.Request
+		err error
+	)
+
+	if dimUpdate.Name == "" || dimUpdate.Value == "" {
```

Comment:
should we check this earlier? in `acceptDimension` or even in `PushKubernetesMetadata`?

Reply:
Moved it for now since it seems to make more sense. When the export supports metadata updates from other sources as well, we may want to refactor.